File: lldb/c3.lua
function string_summary(valobj, internal_dict)
    -- Get the fields from the C3 String struct
    local ptr_obj = valobj:GetChildMemberWithName("ptr")
    local len_obj = valobj:GetChildMemberWithName("len")
    
    if not ptr_obj:IsValid() or not len_obj:IsValid() then
        return "<invalid string>"
    end

    local length = len_obj:GetValueAsUnsigned()
    local address = ptr_obj:GetValueAsUnsigned()

    -- Read the string from memory
    local process = lldb.debugger:GetSelectedTarget():GetProcess()
    local error = lldb.SBError()
    local content = process:ReadMemory(address, length, error)
    
    if error:Success() then
        return '"' .. content .. '"'
    else
        return "<error reading memory>"
    end
end

File: lldb/c3.py
import lldb

def String_Summary(valobj, internal_dict):
    # Get the pointer and length fields from your 'String' struct
    ptr_val = valobj.GetChildMemberWithName("ptr").GetValueAsUnsigned(0)
    len_val = valobj.GetChildMemberWithName("len").GetValueAsUnsigned(0)
    
    if len_val == 0:
        return '""'
    
    # Read exactly 'len_val' bytes from the process memory
    error = lldb.SBError()
    process = valobj.GetProcess()
    data = process.ReadMemory(ptr_val, len_val, error)
    
    if error.Success():
        # Decode as UTF-8 or keep as raw bytes
        return f'"{data.decode("utf-8", "replace")}"'
    else:
        return f'<error reading memory: {error.GetCString()}>'

